레스토랑 조회 구현하기

wkddngus5 opened the pull request:
## 구조

components:  지역과 카테고리 필터를 담은 Form, 레스토랑 list
services: apiService

actions에서는 apiService를 이용해 api 통신 결과를 비동기적으로 dispatch하는 로직

---
## 질문
coverage 테스트를 해보면 actions에서 부족한 부분이 나오는데 apiService 클래스를 모킹해서 테스트해야 할까요?

wkddngus5 committed:
SearchFilterContainer region options 렌더링 테스트 코드 작성

wkddngus5 committed:
지역 목록 store에서 받아와 렌더링

wkddngus5 committed:
category options 렌더링 테스트 코드

wkddngus5 committed:
options 초기화 테스트 코드

wkddngus5 committed:
redux-thunk로 비동기 통신을 통한 옵션 초기화

wkddngus5 committed:
SearchFilter 컴포넌트 옵션 렌더링 테스트

wkddngus5 committed:
SearchFilter 컴포넌트

wkddngus5 committed:
SearchFilter value, onChange 테스트 코드

wkddngus5 committed:
SearchFilter value, onChange 인터페이스 구현

wkddngus5 committed:
SearchFilterContainer changeFilter 테스트

wkddngus5 committed:
SearchFilterContainer onChange 구현

wkddngus5 committed:
레스토랑 리스트 렌더링 테스트

wkddngus5 committed:
레스토랑 리스트 렌더링

wkddngus5 committed:
리팩토링
헤딩 추가
ApiService 추가

wkddngus5 committed:
reducer, ApiService 테스트 코드 작성

wkddngus5 committed:
select 태그를 ul li 태그 조합으로 변경

wkddngus5 committed:
오타 수정

hannut91 commented:
> coverage 테스트를 해보면 actions에서 부족한 부분이 나오는데 apiService 클래스를 모킹해서 테스트해야 할까요?

커버리지 결과를 정확히 봐야 할 것 같습니다. 테스트를 실행하실 때 커버리지를 출력하도록 한 후, 다음 파일을 열면 자세히 볼 수 있습니다.

```bash
npm run test:unit -- --coverage
# 브라우저에서 열기
open ./coverage/lcov-report/index.html
```

hannut91 commented on src/App.jsx:
```diff
@@ -1,5 +1,26 @@
   return (
-    <></>
+    <main>
+      <h1 style={{ display: 'none' }}>레스토랑 검색</h1>
```
컴포넌트로 추출하는 것은 어떨까요?

hannut91 commented on src/components/RestaurantsListContainer/RestaurantsList.jsx:
```diff
@@ -0,0 +1,14 @@
+      <ul>
+        {restaurants.map(({ id, name }) => (
+          <li key={id}>
+            <h3>{name}</h3>
```
여기서는 `h3`을 사용하면 안될 것 같아요. 태그가 없는게 맞을 것 같아요

wkddngus5 replied:
restaurant라는 각각의 article에 대한 타이틀이라고 생각했는데 혹시 사용하면 안될것 같은 이유가 있을까요?

hannut91 replied:
음 그렇군요 ㅎㅎ 저는 리스트 안에 타이틀이 어색하다고 생각했었어요

hannut91 commented on src/store/reducer.test.js:
```diff
@@ -0,0 +1,42 @@
+
+  context('when setRegion is dispatched', () => {
+    it('returns state with new regions', () => {
+      const regions = [{ id: 1, name: '서울' }];
+
+      expect(reducer(undefined, setRegions(regions)).regions).toEqual(regions);
```
```suggestion
      const regions = [{ id: 1, name: '서울' }];
      
      const state = reducer(undefined, setRegions(regions));

      expect(state.regions).toEqual(regions);
```

중간값을 변수로 추출하면 더 이해하기 좋을 것 같아요. 왜냐하면 테스트 이름에 `state`가 들어갔기 때문이에요

wkddngus5 replied:
좋은것 같습니다!

hannut91 commented on src/store/reducer.js:
```diff
@@ -0,0 +1,43 @@
+};
+
+export default function reducer(state = initialState, action) {
+  if (action.type === ACTION_TYPES.SET_REGIONS) {
```
if문이 많아서 복잡해지는 것 같습니다. if문 혹은 switch문 없이, 즉 논리적인 명령문을 사용하지 않고 같은 기능을 구현할 수 있을까요?

wkddngus5 replied:
key-value 쌍을 가진 reducers object에서 action.type이라는 키값을 기반으로 가져오도록 수정했습니다!

hannut91 commented on src/store/actions.js:
```diff
@@ -0,0 +1,66 @@
+    dispatch(setCategories(categories));
+
+    const { filter } = getState();
+    dispatch(setFilter({ ...filter, category: categories[0].id }));
```
왜 `0`인지 이해하기 어려운 것 같아요

wkddngus5 replied:
`DEFAULT_OPTION_INDEX`라는 상수로 추출했습니다.

hannut91 replied:
아 이게 목록을 불러오면 첫 번째 것을 기본으로 선택되게 하고 싶으셨던 거군요. 그렇다면 

```js
    const regions = await apiService.fetchRegions();
    dispatch(setRegions(regions));

    const { filter } = getState();
    dispatch(
      setFilter({ ...filter, region: first(regions).id }),
    );
```

`first`함수같은 것을 사용하거나 혹은

```js
    const regions = await apiService.fetchRegions();
    dispatch(setRegions(regions));

    const { filter } = getState();
    const [region] = regions;
    dispatch(
      setFilter({ ...filter, region: region.id }),
    );
```

혹은

```js
const defaultSelectedRegionId  = (regions) => regions[0].id;

// ...
    const regions = await apiService.fetchRegions();
    dispatch(setRegions(regions));

    const { filter } = getState();
    const regionId = defaultSelectedRegionId(regions);
    dispatch(
      setFilter({ ...filter, region: regionId }),
    );
```

함수로 추출해도 괜찮을 것 같아요. 단순히 상수를 도입한다고 해서 더 잘 이해할 수 있는 경우는 아닌 것 같아요. 오히려 매직 리터럴인 0이 더 좋고, 이것도 싫다면 위와 같은 선택지들도 있을 것 같아요

wkddngus5 replied:
감사합니다! first 함수와 해설강의 보면서 get 함수도 테스트 코드와 함께 만들어서 적용시켜보았습니다:)

hannut91 commented on src/services/api/ApiService.test.js:
```diff
@@ -0,0 +1,66 @@
+  const apiService = new ApiService(host, logger);
+
+  context('with api success', () => {
+    global.fetch = jest.fn(() => Promise.resolve({
+      json: () => Promise.resolve({}),
+    }));
```
```suggestion
    global.fetch = jest.fn(async () => ({
      json: async () => ({}),
    }));
```

hannut91 commented on src/services/api/ApiService.test.js:
```diff
@@ -0,0 +1,66 @@
+  const logger = { error: jest.fn() };
+  const apiService = new ApiService(host, logger);
+
+  context('with api success', () => {
```
테스트 대상이 분명하게 드러나지 않는 것 같아요. 성공한 케이스를 모두 하나로 묶은 것 같은데, 각 API 마다 테스트를 작성하는게 좋을 것 같아요

wkddngus5 replied:
fetch 함수를 context별로 모킹해주기 위해 다음과 같이 컨텍스트로 나누었었는데 우선 에러 부분 테스트코드를 제거하고 context에서 추출했습니다~

hannut91 commented on src/services/api/ApiService.test.js:
```diff
@@ -0,0 +1,66 @@
+    });
+  });
+
+  context('with api error', () => {
```
어떤 것을 드러내고 싶으셨는지 잘 모르겠어요. 이 API의 사용법을 익히기 위해 이 테스트 코드를 보았는데, 어떤 경우에 에러가 나고, 그래서 어떻게 처리해야하는지 전혀 모르겠어요

wkddngus5 replied:
ApiService(describe) - with api error(context) - calls logger.error in fetchRegions(it)
말 그대로 `ApiService에서 api 통신 error가 발생하면 logger에서 error를 출력한다`라는 부분을 테스트하고 싶었습니다.
fetch api를 통한 통신 과정에서 에러가 발생할 수도 있기 때문입니다.
현재 ApiService에서는 비동기 통신을 할 때 try-catch를 통해 통신 과정에서 에러가 발생하면 로그를 찍도록 구현했습니다!

hannut91 commented on src/components/SearchFilterContainer/index.test.jsx:
```diff
@@ -0,0 +1,74 @@
+
+jest.mock('react-redux');
+
+describe('SearchFilterContainer', () => {
```
0, 1 Maglic literal이 계속 사용되고 있는데 무엇을 나타내는지 이해하기 어려운 것 같아요

wkddngus5 replied:
변수로 추출했습니다~!

hannut91 commented on src/components/SearchFilterContainer/SearchFilter.jsx:
```diff
@@ -0,0 +1,34 @@
+export default function SearchFilter({ options: { regions, categories }, value = {}, onChange }) {
```
복잡성이 증가한 것 같네요. 왜냐하면 두 가지 일을 하고있기 때문이에요. 관심사를 적절히 분리해주면 좋을 것 같아요

wkddngus5 replied:
지역을 선택하는 SelectRegion과 카테고리를 선택하는 SelectCategory 두개의 컴포넌트로 분리했는데 이 방향이 맞을까요?
또 분리하면서 든 생각인데 테스트 커버리지를 위해서 컴포넌트가 이런식으로 쪼개질때마다 테스트 코드도 함께 대규모 수정이 발생하는게 과연 맞을까? 하는 의문도 드는데 어떻게 생각하시나요?

hannut91 replied:
테스트코드도 코드이기 때문에 지속적인 리팩터링을 해주어야 해요. 테스트 코드를 작성할 때, 서로 관련이 없는 코드들이 반복된다면, 의미있는 단위로 테스트를 나누어야 한다는 신호로 받아들여야 합니다.

wkddngus5 replied:
리팩토링 내성이라는 개념에 너무 신경쓰니까 오히려 수정에 대해 거부감을 가졌었던것 같네요! 감사합니다:)

wkddngus5 committed:
리뷰 반영
Title 컴포넌트 분리
h3 태그 제거
magic number 제거
ApiService 에러 핸들링 테스트 제거
reducer에 if문 제거

wkddngus5 committed:
리뷰 반영
SearchFilter에 Select 컴포넌트 분리

hannut91 commented on src/components/SearchFilterContainer/SearchFilter.test.jsx:
```diff
@@ -0,0 +1,83 @@
+      },
+    });
+
+    expect(getByText(`${regions[1].name}(V)`)).toBeInTheDocument();
+    expect(getByText(`${categories[1].name}(V)`)).toBeInTheDocument();
+  });
+
+  context('when region is changed', () => {
+    it('calls onChange', () => {
+      const { getByText } = renderComponent({
+        options: {
+          regions,
+          categories,
+        },
+        value: {
+          region: regions[0].id,
+          category: categories[0].id,
+        },
+      });
+
+      fireEvent.click(getByText(regions[1].name));
+      expect(handleChange).toBeCalledWith({ region: regions[1].id, category: categories[0].id });
+    });
+  });
+
+  context('when category is changed', () => {
+    it('calls onChange', () => {
+      const { getByText } = renderComponent({
+        options: { regions, categories },
+        value: {
+          region: regions[0].id,
+          category: categories[0].id,
+        },
+      });
+
+      fireEvent.click(getByText(categories[1].name));
+      expect(handleChange).toBeCalledWith({ region: regions[0].id, category: categories[1].id });
+    });
+  });
+});
```
```jsx
describe('SearchFilter', () => {
  const options = {
    regions,
    categories,
  };

  const handleChange = jest.fn();

  given('value', () => ({
    region: undefined,
    category: undefined,
  }));

  const renderComponent = () => render((
    <SearchFilter 
      options={options}
      value={given.value}
      onChange={handleChange} 
    />
  ))

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders regions and categories', () => {
    const { getByText } = renderComponent();

    expect(getByText('서울')).toBeInTheDocument();
    expect(getByText('한식')).toBeInTheDocument();
  });

  context('when region and category selected', () => {
    given('value', () => ({
      region: regions[0].id,
      category: categories[0].id,
    }));

    it('renders selected option value', () => {
      const { getByText } = renderComponent();
  
      expect(getByText(`${regions[0].name}(V)`)).toBeInTheDocument();
      expect(getByText(`${categories[0].name}(V)`)).toBeInTheDocument();
    });

    describe('Click region', () => {
      it('calls onChange handler', () => {
        const { getByText } = renderComponent();
      
        fireEvent.click(getByText(regions[1].name));

        expect(handleChange)
          .toBeCalledWith({ region: regions[1].id, category: categories[0].id });
      });
    });

    describe('Click category', () => {
      it('calls onChange handler', () => {
        const { getByText } = renderComponent();
  
        fireEvent.click(getByText(categories[1].name));
        expect(handleChange).toBeCalledWith({ region: regions[0].id, category: categories[1].id });
      });
    });
  });
});
```

given2를 이용해서 테스트를 작성해 보았습니다. 참고해서 적용해 보시면 좋을 것 같아요

See also
  - https://github.com/tatyshev/given2

wkddngus5 replied:
감사합니다! 이전 수강생 동우님께 given2 배우기는 했는데 다음주 과제부터는 적극적으로 사용해보겠습니다:)

wkddngus5 committed:
first method

hannut91 commented:
과제풀이를 보고 꼭 복습해 보세요!

wkddngus5 committed:
_.get 메서드

wkddngus5 committed:
App.jsx에서 렌더링과 useEffect 테스트

wkddngus5 committed:
리뷰 반영
검색 필터 테스트코드 구조 수정

wkddngus5 committed:
ApiService Http 통신 에러 핸들링 테스트 코드 복구

wkddngus5 committed:
action.js
비동기 action 테스트 코드 추가
매직 넘버 _.fist함수로 대체

wkddngus5 committed:
필요한 모킹 처리

hannut91 commented:
> 때로는 describe 내부에 다시 describe가 들어가거나, describe의 하위 개념이라고 생각했던 context 내부에 상위 개념이라고 생각했던 describe가 들어갈 수도 있다는 것을 배웠고, 아마 실무에서는 구성원들간의 충분한 토론과 합의를 통해서 적절한 컨벤션을 채택하지 않을까라는 생각이 들었습니다:)

저도 이런 부분들에 대해서 고민이 많은데, 너무 많이 중첩되면 테스트에 대해서 무언가 잘못되었다는 신호로 받아들이시면 좋습니다. 저도 2번 까지는 중첩을 해서 작성하긴 하는데, 그 이상은 조심하고 있습니다.

이번 한 주도 고생 많으셨습니다!
